src/vmm/src/builder.rs

```diff
@@ -324,21 +269,128 @@ pub fn build_microvm(
+        .restore_state(&microvm_state.vm_state)
+        .map_err(MicrovmStateError::RestoreVmState)
+        .map_err(RestoreMicrovmState)?;
+
+    // Restore devices states.
+    let mmio_ctor_args = MMIODevManagerConstructorArgs {
+        mem: guest_memory,
+        vm: vmm.vm.fd(),
+        event_manager,
+    };
+    vmm.mmio_device_manager =
+        MMIODeviceManager::restore(mmio_ctor_args, &microvm_state.device_states)
+            .map_err(MicrovmStateError::RestoreDevices)
+            .map_err(RestoreMicrovmState)?;
+
+    // Move vcpus to their own threads and start their state machine in the 'Paused' state.
+    vmm.start_vcpus(vcpus, seccomp_filter)
+        .map_err(StartMicrovmError::Internal)?;
+
+    // Restore vcpus kvm state.
+    vmm.restore_vcpu_states(microvm_state.vcpu_states)
+        .map_err(RestoreMicrovmState)?;
+
+    // Load seccomp filters for the VMM thread.
+    // Execution panics if filters cannot be loaded, use --seccomp-level=0 if skipping filters
+    // altogether is the desired behaviour.
+    SeccompFilter::apply(seccomp_filter.to_vec())
+        .map_err(Error::SeccompFilters)
+        .map_err(StartMicrovmError::Internal)?;
+
+    let vmm = Arc::new(Mutex::new(vmm));
+    event_manager
+        .add_subscriber(vmm.clone())
+        .map_err(StartMicrovmError::RegisterEvent)?;
+
+    Ok(vmm)
+}
+
+/// Builds and starts a microVM based on the current Firecracker VmResources configuration.
+///
+/// This is the default build recipe, one could build other microVM flavors by using the
+/// independent functions in this module instead of calling this recipe.
+///
+/// An `Arc` reference of the built `Vmm` is also plugged in the `EventManager`, while another
+/// is returned.
+pub fn build_microvm_for_boot(
+    vm_resources: &super::resources::VmResources,
+    event_manager: &mut EventManager,
+    seccomp_filter: BpfProgramRef,
+) -> std::result::Result<Arc<Mutex<Vmm>>, StartMicrovmError> {
+    use self::StartMicrovmError::*;
+    let boot_config = vm_resources.boot_source().ok_or(MissingKernelConfig)?;
+
+    let guest_memory = create_guest_memory(
+        vm_resources
+            .vm_config()
+            .mem_size_mib
+            .ok_or(MissingMemSizeConfig)?,
+    )?;
+    let vcpu_config = vm_resources.vcpu_config();
+    let track_dirty_pages = vm_resources.track_dirty_pages();
+    let entry_addr = load_kernel(boot_config, &guest_memory)?;
+    let initrd = load_initrd_from_config(boot_config, &guest_memory)?;
+    // Clone the command-line so that a failed boot doesn't pollute the original.
+    #[allow(unused_mut)]
+    let mut boot_cmdline = boot_config.cmdline.clone();
+
+    let (mut vmm, mut vcpus) = build_microvm_common(
+        event_manager,
+        guest_memory,
+        track_dirty_pages,
+        vcpu_config.vcpu_count,
+    )?;
+
     attach_block_devices(
         &mut vmm,
         &mut boot_cmdline,
-        &vm_resources.block,
+        vm_resources.block.list.iter(),
```

Comment:
> Having an iterator as function parameter doesn't look like something very intuitive or idiomatic

Why not?

--------------

This change is ~required~ _really-nice-to-have_ to break the tight coupling between this generic microVM building code from the Firecracker-specific `BlockBuilder`/`NetBuilder`/`VsockBuilder`.

Using a generic iterator here allows calling this function from any specific code as long as that code/class/object can provide an iterator over the actual devices.

Reply:
> > Having an iterator as function parameter doesn't look like something very intuitive or idiomatic
> 
> Why not?
> 
Well, it's not the first thing you're thinking about when you want to write a function that does something with the elements of a specific collection :D. Anyway, my observation is indeed pretty subjective, but I find a function which takes a reference to a collection of `<T>` instead of an iterator over `<T>` elements a little bit more readable and easier to understand, also there's no need to add explicit lifetimes which in general affect readability too.

> This change is ~required~ _really-nice-to-have_ to break the tight coupling between this generic microVM building code from the Firecracker-specific `BlockBuilder`/`NetBuilder`/`VsockBuilder`.
> 
> Using a generic iterator here allows calling this function from any specific code as long as that code/class/object can provide an iterator over the actual devices.

I agree it decouples things a little, the fact that we're using for example a `VecDeque` in `BlockBuilder` is pretty specific, but at this moment those functions are used only in the `BlockBuilder`, `NetBuilder` contexts, so I don't know if this change is very useful now. Maybe it would be a good idea to refactor a bit `insert_block_devices()` and `insert_net_device()` from tests to not use `BlockBuilder`, `NetBuilder` anymore and this way your change will be even more justified.

Reply:
In tests we're using `BlockBuilder`/`NetBuilder`/`VsockBuilder` to get free device building - so we don't duplicate code to build the device.

I, too, don't like how the dependency graph is a mess when looking at tests.
We should address this as part of https://github.com/firecracker-microvm/firecracker/issues/1713.